### PR TITLE
fix autocompletion for rue de la loire

### PR DESCRIPTION
**Pull request by xlqian**
fix autocompletion for "rue de la loire"

http://jira.canaltp.fr/browse/NAVITIAII-2096

### CODE REVIEW

File: source/autocomplete/autocomplete.h [hunk 1]
```diff
@@ -304,13 +304,12 @@ struct Autocomplete
...
-        return input;
+    void sort_and_truncate_by_score(std::vector<fl_quality>& input, size_t nbmax) const {
+        sort_and_truncate(input, nbmax, [](const fl_quality& a, const fl_quality& b){return a.score > b.score;});
     }
 
-    std::vector<fl_quality> sort_and_truncate_by_quality(std::vector<fl_quality> input, size_t nbmax) const {
-        sort_and_truncate(input, nbmax, [](fl_quality a, fl_quality b){return a.quality > b.quality;});
+    std::vector<fl_quality> sort_and_truncate_by_quality(std::vector<fl_quality>& input, size_t nbmax) const {
+        sort_and_truncate(input, nbmax, [](const fl_quality& a, const fl_quality& b){return a.quality > b.quality;});
         return input;
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.h:313]: Why returning a copy of the vector if you take it by mut ref?

Comment (xlqian) [inline source/autocomplete/autocomplete.h:313]: as you can see, it's not my code... I changed the one used by my part but not this one

File: source/autocomplete/autocomplete.cpp [hunk 2]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
     }
 }
 
+template<class T>
+std::vector<typename Autocomplete<T>::fl_quality>
+Autocomplete<T>::compute_vec_quality(const std::string& str ,
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: spacing

File: source/autocomplete/autocomplete.cpp [hunk 3]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+                                     const std::vector<T>& index_result,
+                                     const navitia::georef::GeoRef& geo_ref,
+                                     std::function<bool(T)> keep_element,
+                                     int wordLength) const{
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: spacing

File: source/autocomplete/autocomplete.cpp [hunk 4]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+    std::vector<fl_quality> vec_quality;
+    std::vector<std::string> tokens_req;
+    {
+        auto tmp_str = strip_accents_and_lower(str);
+        boost::split(tokens_req, tmp_str, boost::is_any_of(" "));
+    }
+    for(auto i : index_result){
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: spacing

File: source/autocomplete/autocomplete.cpp [hunk 5]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+        if(keep_element(i)) {
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: spacing

File: source/autocomplete/autocomplete.cpp [hunk 6]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+            quality.idx = i;
+            quality.nb_found = word_quality_list.at(quality.idx).word_count;
+            quality.word_len = wordLength;
+            quality.score = word_quality_list.at(quality.idx).score;
+            quality.quality = 100;
+
+            // whole_way_name = road name + city name
+            auto whole_way_name = geo_ref.ways[i]->name;
+            for (const auto* admin: geo_ref.ways[i]->admin_list) {
+                if(admin && admin->level == 8) {
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: spacing

File: source/autocomplete/autocomplete.cpp [hunk 7]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+                    whole_way_name += (" " + admin->name);
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp:185]: we don't have a label for that?

File: source/autocomplete/autocomplete.cpp [hunk 8]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+                }
+            }
+            whole_way_name = strip_accents_and_lower(whole_way_name);
+            boost::algorithm::replace_all(whole_way_name, "-", " ");
+
+            boost::tokenizer<> tokens_candidate(whole_way_name);
+            auto it_req = std::begin(tokens_req);
+            auto it_can = std::begin(tokens_candidate);
+            for(; it_req != std::end(tokens_req) && it_can != std::end(tokens_candidate); ++it_req, ++it_can) {
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: spacing

File: source/autocomplete/autocomplete.cpp [hunk 9]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+            }
+            vec_quality.emplace_back(std::move(quality));
+        }
+    }
+    return vec_quality;
+}
+
+template<class T>
+std::vector<typename Autocomplete<T>::fl_quality>
+Autocomplete<T>::find_complete_way(const std::string & str,
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: spacing

File: source/autocomplete/autocomplete.cpp [hunk 10]
```diff
@@ -155,4 +156,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+                }
+            }
+            whole_way_name = strip_accents_and_lower(whole_way_name);
+            boost::algorithm::replace_all(whole_way_name, "-", " ");
+
+            boost::tokenizer<> tokens_candidate(whole_way_name);
+            auto it_req = std::begin(tokens_req);
+            auto it_can = std::begin(tokens_candidate);
+            for(; it_req != std::end(tokens_req) && it_can != std::end(tokens_candidate); ++it_req, ++it_can) {
+                quality.score += (*it_can).find(*it_req) != std::string::npos;
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp:195]: thus, we add one to the score if the token are exactly the same?

File: source/autocomplete/autocomplete.cpp (hunk 7 above)
Comment (xlqian) [inline source/autocomplete/autocomplete.cpp:185]: get_label will return a string with `()`

File: source/autocomplete/autocomplete.cpp (hunk 10 above)
Comment (xlqian) [inline source/autocomplete/autocomplete.cpp:195]: not exact match,  if the request word is `sai seb`, the score will have two additional points if the candidate is `saint sebastien`

File: source/autocomplete/autocomplete.cpp [hunk 11]
```diff
@@ -158,19 +157,19 @@ void Autocomplete<T>::compute_score(type::PT_Data &pt_data, georef::GeoRef &geor
...
     std::vector<fl_quality> vec_quality;
     std::vector<std::string> tokens_req;
     {
         auto tmp_str = strip_accents_and_lower(str);
         boost::split(tokens_req, tmp_str, boost::is_any_of(" "));
     }
-    for(auto i : index_result){
-        if(keep_element(i)) {
+    for ( auto i : index_result ) {
+        if ( keep_element(i) ) {
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: Noooo!!!!!!!

File: source/autocomplete/autocomplete.h [hunk 12]
```diff
@@ -341,10 +340,23 @@ struct Autocomplete
...
     }
 
 
+    std::vector<fl_quality> compute_vec_quality(const std::string& str,
+                                                const std::vector<T>& index_result,
+                                                const navitia::georef::GeoRef& geo_ref,
+                                                std::function<bool(T)> keep_element,
+                                                int word_length) const;
+
+    std::vector<fl_quality> find_complete_way(const std::string & str,
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.h]: BWAAAAA!

File: source/autocomplete/autocomplete.cpp [hunk 13]
```diff
@@ -155,4 +155,69 @@ void Autocomplete<type::idx_t>::compute_score(type::PT_Data &pt_data, georef::Ge
...
+            quality.idx = i;
+            quality.nb_found = word_quality_list.at(quality.idx).word_count;
+            quality.word_len = wordLength;
+            quality.score = word_quality_list.at(quality.idx).score;
+            quality.quality = 100;
+
+            // whole_way_name = road name + "(" +  city name + ")"
+            auto whole_way_name = geo_ref.ways[i]->name;
+            for (const auto* admin: geo_ref.ways[i]->admin_list) {
+                if(admin && admin->level == 8) {
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: Pompompom...

**Pull request merged**